### PR TITLE
Add component completion

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
This pull request collects components in script part and feed components to template completion.

Prop name completion
screenshot from 2017-08-20 01-11-49

Component name completion
screenshot from 2017-08-20 01-11-30

Still needs some polishing and tests.

### CODE REVIEW

**Comment by HerringtonDarkholme** [comment]
Test and polish has been done. This pull request is ready to go!

**Comment by octref** [comment]
I was busy preparing for a talk in today's Vue NYC meetup.
This seems to work fine and I'll demo it (thanks a lot!).
But I'll take a deeper look tonight and see if there is any issues. We'll have this feature for the next release.

**Comment by HerringtonDarkholme** [comment]
Ping @octref

File: server/src/modes/script/findComponents.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import * as ts from 'typescript';
+
+export interface ComponentInfo {
+  name: string;
+  props?: string[];
+}
+
+export function findComponents(service: ts.LanguageService, fileFsPath: string) {
```
Comment (octref) [inline server/src/modes/script/findComponents.ts]: This actually returns `ComponentInfo[]`. So maybe naming it as `findCompInfos`. Also feel explicitly writing the return time helps with code readability.

File: server/src/modes/script/findComponents.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  const sourceFile = program.getSourceFile(fileFsPath);
+  const importStmt = sourceFile.statements.filter(st => st.kind === ts.SyntaxKind.ExportAssignment);
+  if (importStmt.length === 0) {
+    return [];
+  }
+  const maybeComp = (importStmt[0] as ts.ExportAssignment).expression;
+  if (maybeComp.kind !== ts.SyntaxKind.CallExpression) {
+    return [];
+  }
+  const instance = maybeComp as ts.CallExpression;
```
Comment (octref) [inline server/src/modes/script/findComponents.ts]: instance isn't used again, so maybe not declaring it (combine line 19-20).

File: server/src/modes/script/findComponents.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+export interface ComponentInfo {
+  name: string;
+  props?: string[];
+}
+
+export function findComponents(service: ts.LanguageService, fileFsPath: string) {
+  const program = service.getProgram();
+  const sourceFile = program.getSourceFile(fileFsPath);
+  const importStmt = sourceFile.statements.filter(st => st.kind === ts.SyntaxKind.ExportAssignment);
```
Comment (octref) [inline server/src/modes/script/findComponents.ts]: You are filtering out Export statements, why call it `importStmt`?

File: server/src/modes/script/findComponents.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  if (importStmt.length === 0) {
+    return [];
+  }
+  const maybeComp = (importStmt[0] as ts.ExportAssignment).expression;
```
Comment (octref) [inline server/src/modes/script/findComponents.ts]: `maybeComp` sounds weird, maybe `potentialComp` or `possibleComp`?

File: server/src/modes/script/findComponents.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  const props = getPropertyTypeOfType(compType, 'props', checker);
+  if (!props) {
+    return info;
+  }
+  info.props = checker.getPropertiesOfType(props).map(s => s.name);
+  return info;
+}
+
+function getArrayProp(compType: ts.Type, checker: ts.TypeChecker) {
```
Comment (octref) [inline server/src/modes/script/findComponents.ts]: `getPropArray` or `getProps`?

File: server/src/modes/script/findComponents.ts (hunk 1 above)
Comment (HerringtonDarkholme) [reply server/src/modes/script/findComponents.ts]: let's first keep the name. I think this code will soon be changed. More features like finding components definition will be added so I guess the function name will need changed.

File: server/src/modes/script/findComponents.ts (hunk 5 above)
Comment (HerringtonDarkholme) [reply server/src/modes/script/findComponents.ts]: the variable `propArray` in l38 is confusing.

This function name can be getArrayProps. 
The notation implies how props are defined, say in array literal, not return type.

**Comment by octref** [comment]
Thanks!

**Pull request merged**